server/routes/routerRequestHandlers/retrieveCuriousFunctionalSkills.ts

```diff
@@ -0,0 +1,21 @@
+import { NextFunction, Request, RequestHandler, Response } from 'express'
+import { CuriousService } from '../../services'
+import asyncMiddleware from '../../middleware/asyncMiddleware'
+
+/**
+ *  Middleware function that returns a Request handler function to look up the prisoner's functional skills from Curious
+ */
+const retrieveCuriousFunctionalSkills = (curiousService: CuriousService): RequestHandler => {
+  return asyncMiddleware(async (req: Request, res: Response, next: NextFunction) => {
+    const { prisonNumber } = req.params
+
+    // Lookup the prisoners functional skills and store in res.locals
+    res.locals.prisonerFunctionalSkills = await curiousService.getPrisonerFunctionalSkills(
```

Comment:
I assume if curios is unavailable we fail to load the whole page?

Reply:
No, actually this is something we put a lot of time and effort into preventing TBH 😁 

The thinking is that a given UI page might need to call a handful of APIs (prisoner-search to get the basic prisoner data, our API to get prisoner goals, curious to get functional skills, curious (again!) to get in-prison courses; etc etc). Each page will have it's own requirements for data from x number of APIs.

For some on the page it might be crucial that we get a response. For example the Prisoner List page, where we call prisoner-search to get the list of prisoners. If that fails, well, the whole page is essentially non-operational. (We have custom error pages, so the user is not left with a white screen of death or a pointless 500 page)

Anyway, for other pages, data from a given API call might not be crucial. It is perhaps supplementary data, and whether its there or not does not impact the operation of the page or core service. For example, on the prisoner overview page we display the prisoner goals, but also display Functional Skills and In-Prison Courses from Curious. If that data isnt there (ie. the Curious API call(s) fail) then the page itself is still operational and we shouldnt block the user from doing their task.

The way we do that is in the service layer.
Basically the service classes call the given REST API, but do not return the API's native response type/data. Instead we map it into a view model before returning it out from the service method.
The service methods (should) always return a view model object, and (should) never return the type from the REST API (I put "should" in brackets - we might have a couple of instances where that is not true! 😬 😢 )

Anyway, the view model object has a property called `problemRetrievingData` - eg:
```
  /**
   * A prisoner's Functional Skills, which is made up of a collection of Assessments.
   */
  export interface FunctionalSkills {
    problemRetrievingData: boolean
    assessments: Array<Assessment>
    prisonNumber: string
  }
```
And the service method has a try/catch that will map the response data from the API into this view model object, ensuring that `problemRetrievingData` is set correctly:
```
  async getPrisonerFunctionalSkills(prisonNumber: string, username: string): Promise<FunctionalSkills> {
    const systemToken = await this.hmppsAuthClient.getSystemClientToken(username)

    try {
      const learnerProfiles = await this.getLearnerProfile(prisonNumber, systemToken)
      return toFunctionalSkills(learnerProfiles, prisonNumber)
    } catch (error) {
      logger.error(`Error retrieving functional skills data from Curious: ${JSON.stringify(error)}`)
      return { problemRetrievingData: true } as FunctionalSkills
    }
  }
```

Lastly, in the nunjucks view layer, we have some simple conditional logic on the `problemRetrievingData` property. EG:
```
<div class="govuk-summary-card app-functional-skills-sidebar">
  <div class="govuk-summary-card__title-wrapper app-summary-card__title-wrapper">
    <h2 class="govuk-summary-card__title govuk-!-font-size-24">Current functional skills assessment scores</h2>
  </div>
  <div class="govuk-summary-card__content">

    {% if not functionalSkills.problemRetrievingData %}
      <table class="govuk-table govuk-!-margin-bottom-0" id="functional-skills-sidebar-table">
        <caption class="govuk-table__caption govuk-visually-hidden">Current functional skills and assessment information</caption>
        <thead class="govuk-table__head govuk-visually-hidden">
          <tr class="govuk-table__row">
            <th scope="col" class="govuk-table__header">Skill and assessment date</th>
            <th scope="col" class="govuk-table__header">Assessment level</th>
          </tr>
        </thead>
        <tbody class="govuk-table__body">
          {% for skill in functionalSkills.assessments %}
          <tr class="govuk-table__row">
            <th scope="row" class="govuk-table__header">
              {{ skill.type | formatFunctionalSkillType }}
              {% if skill.assessmentDate %}
              <br>
              <span class="govuk-body-s">Assessed on {{ skill.assessmentDate | formatDate('D MMMM YYYY') }}</span>
              {% endif %}
            </th>
            <td class="govuk-table__cell">
              {% if skill.grade %}{{ skill.grade }}{% else %}No functional skill assessment scores recorded in Curious{% endif %}
            </td>
          </tr>
          {% endfor %}
        </tbody>
      </table>
      <p class="govuk-body govuk-!-margin-top-2 govuk-!-display-none-print">
        <a class="govuk-link" href="../functional-skills" data-qa="view-all-functional-skills-button">View all <span class="govuk-visually-hidden">functional skills</span></a>
      </p>
    {% else %}
      <h3 class="govuk-heading-s" data-qa="functional-skills-sidebar-error-heading">We cannot show these details from Curious right now</h3>
      <p class="govuk-body">Reload the page or try again later. Other parts of this service may still be available.</p>
    {% endif %}
  </div>
</div>
```

Reply:
IE. like this:
![Screenshot 2024-06-14 at 08 49 43](https://github.com/ministryofjustice/hmpps-education-and-work-plan-ui/assets/94835226/bbbb9286-0e67-40e9-ae40-452dcd7a9062)